src/v/kafka/server/offset_translator.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+#include "kafka/types.h"
+#include "model/fundamental.h"
+#include "raft/configuration_manager.h"
+
+namespace kafka {
+/**
+ * Offset translator perfroms conversion between kafka offsets, which doesn't
+ * include raft configuration batches and log offsets which includes
+ */
+class offset_translator {
+public:
+    explicit offset_translator(const raft::configuration_manager& config_mgr)
+      : _cfg_mgr(config_mgr) {}
+
+    model::offset to_kafka_offset(model::offset o) const {
+        auto d = delta(o);
+        return model::offset(o - d);
+    }
+
+    model::offset from_kafka_offset(
+      model::offset o, model::offset hint = model::offset{}) const {
+        if (o == model::offset::max()) {
+            return o;
+        }
+        model::offset kafka_offset(o);
+        auto current = std::max(hint, kafka_offset);
+        auto next_cfg_it = _cfg_mgr.lower_bound(current);
+        // fast exit, current offset is larger than last config batch
+        // offset, just add delta
+        if (next_cfg_it == _cfg_mgr.end()) {
+            --next_cfg_it;
+            return model::offset(o) + next_cfg_it->second.idx();
+        }
+
+        auto delta = model::offset(next_cfg_it->second.idx() - 1);
+
+        while (current - delta != kafka_offset
+               || current == next_cfg_it->first) {
+            auto& [next_cfg_offset, _] = *next_cfg_it;
+            // largest kafka offset that is achievable before reaching next
+            // control batch
+            const auto max_cfg_ko = next_cfg_offset - delta;
+            if (max_cfg_ko > o) {
+                // offset is between current and next configuration batch,
+                // it is enough to add current delta
+                return kafka_offset + delta;
+            }
+            // skip to the next configuration batch
+            auto prev = next_cfg_it;
+            ++next_cfg_it;
+            delta += model::offset(std::distance(prev, next_cfg_it));
+
+            if (next_cfg_it == _cfg_mgr.end()) {
+                return kafka_offset + delta;
+            }
+        }
```

Comment:
I need to think about this some more. It feels like there's enough information to advance `next_cfg_it` by more than one each time, maybe related to `(o - max_cfg_ko)`.

Reply:
I was considering this, we can implement a binary search here, then we could advance it by more than one. Usually control batches are sparse so it would perform single lookup for most scenarios. I have also optimization in mind to store last translation result and pass it as a hint. this way it would always be exactly one lookup.